spec.bs

```diff
@@ -2664,6 +2695,82 @@ To <dfn>fetch trusted signals</dfn> given a [=URL=] |url|, an [=origin=] |script
   1. Return « |signals|, |perInterestGroupData|, |dataVersion| ».
 </div>
 
+<div algorithm>
+To <dfn>fetch trusted key value signals</dfn> given a [=URL=] |url|, a [=byte sequence=] |body|, a [=string=]
+|context|, an [=origin=] |scriptOrigin|, a [=policy container=] |policyContainer|, an |indexMap|, and a
+[=boolean=] |isBiddingSignal|:
+
+1. Let |request| be a new [=request=] with the following properties:
+  :   [=request/URL=]
+  ::  |url|
+  :   [=request/method=]
+  ::  `POST`
+  :   [=request/body=]
+  ::  |body|
+  :   [=request/origin=]
+  ::  |scriptOrigin|
+  :   [=request/header list=]
+  ::  «`Content-Type`: `message/ad-auction-trusted-signals-request`»
+  ::  «`Accept`: `message/ad-auction-trusted-signals-response`»
+  :   [=request/client=]
+  ::  `null`
+  :   [=request/mode=]
+  ::  "`cors`"
+  :   [=request/referrer=]
+  ::  "`no-referrer`"
+  :   [=request/credentials mode=]
+  ::  "`omit`"
+  :   [=request/redirect mode=]
+  ::  "`error`"
+  :   [=request/service-workers mode=]
+  ::  "`none`"
+  :   [=request/policy container=]
+  ::  A new [=policy container=] whose [=policy container/IP address space=] is |policyContainer|'s
+    [=policy container/IP address space=]
+
+1. Let |signals| be null.
+1. Let |dataVersion| be an empty [=ordered map=], whose [=map/keys=] are [=strings=] and [=map/values=] are integers.
+1. Let |perInterestGroupData| be an [=ordered map=] whose [=map/keys=] are [=interest group/name=] [=strings=]
+  and whose [=map/values=] are [=bidding signals per interest group data=].
+1. [=Fetch=] |request| with [=fetch/useParallelQueue=] set to true, and
+  [=fetch/processResponseConsumeBody=] set to the following steps given a [=response=] |response|
+  and null, failure, or a [=byte sequence=] |responseBody|:
+  1. If [=validate fetching response=] with |response|, |responseBody| and
+    "`message/ad-auction-trusted-signals-response`" returns false, set |signals| to failure and return.
+  1. Let |resultList| be the result of deserializing |responseBody| using |context|. The
+    deserialization method may follow that described in
+    [Section 2.3.6 of the Protected Audience Key Value Services](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#name-parsing-a-response).
+  1. [=list/For each=] |result| in |resultList|:
+    1. If |isBiddingSignal| is true:
+      1. [=map/For each=] |name| → |value| in |result|["interestGroupNames"]:
+        1. If |indexMap|[|name|] does not equal to |result|["index"], set |signals| to failure and return.
+        1. [=map/Set=] |perInterestGroupData|[|name|] to |value|.
+        1. If |result| [=map/contain=] "dataVersion":
+          1. If |result|["dataVersion"] is not an integer, or is less than 0 or more than

```

Comment:
nit: dataVersion is a field of |result|, so you can check its value outside of this for loop, right under "1. [=list/For each=] |result| in |resultList|:", so that you don't need to do the check for each name, and don't need to duplicate the check for buyer/seller. Then you can just set |dataVersion|[name] here without this check step.